packages/gatsby-source-shopify/src/gatsby-node.js

```diff
@@ -25,20 +27,54 @@ import {
 } from "./queries"
 
 export const sourceNodes = async (
-  { actions: { createNode, touchNode }, createNodeId, store, cache },
+  { actions: { createNode, touchNode }, getNodes, createNodeId, store, cache },
   { shopName, accessToken, verbose = true, paginationSize = 250 }
 ) => {
-  const client = createClient(shopName, accessToken)
+  // Arguments used for file node creation.
+  const imageArgs = { createNode, createNodeId, touchNode, store, cache }
 
   // Convenience function to namespace console messages.
   const formatMsg = msg =>
     chalk`\n{blue gatsby-source-shopify/${shopName}} ${msg}`
 
   try {
-    console.log(formatMsg(`starting to fetch data from Shopify`))
+    const online = await isOnline()
+    // if the user sets the GATSBY_SHOPIFY_OFFLINE environment variable
+    // force a fall-back to cached nodes, but never in production
+    const preferCache =
+      process.env.GATSBY_SHOPIFY_OFFLINE === `true` &&
+      process.env.NODE_ENV !== `production`
+
+    if (!online || preferCache) {
+      console.log(chalk`{bold.red Using offline cache for Shopify} ⚠️`)
+      console.log(
+        chalk`{bold.red Cache may be invalidated if you edit package.json, gatsby-node.js or gatsby-config.js files}`
+      )
+
+      // get all nodes created by this source plugin
+      getNodes()
+        .filter(n => n.internal.owner === `gatsby-source-shopify`)
+        .forEach(node => {
+          // touch node to persist in cache
+          touchNode({ nodeId: node.id })
+          // check for image nodes and touch them to persist in cache
+          if (node.image) {
+            touchNode({ nodeId: node.image.localFile___NODE })
+          }
+          if (node.images && node.images.edges) {
+            node.images.edges.map(edge => {
+              touchNode({ nodeId: edge.node.id })
+            })
+          }
```

Comment:
This is fragile. Ideally we don't target `image` and `images`/`images.edges` here and we have more abstract way of traversing node to touch linked nodes. This might break later on when we introduce changes to other parts of shopify plugin code and this can go unnoticed

Reply:
Can you expand a little on what you mean by "a more abstract way of traversing node to touch linked nodes"? What other linked nodes are there I should be aware of? Any tips on how I can implement this in a way which ensures all linked nodes are touched?

Reply:
> What other linked nodes are there I should be aware of? 

Yeah, the point is that we don't know and we should not make any assumptions that we know. It's very easy to desync other parts of code with feature like this that might not be used very often (and easy to break them).

> Any tips on how I can implement this in a way which ensures all linked nodes are touched?

I think if we do deep traverse of node, we can look for fields that end in `___NODE` (that's how we link nodes) and touch nodes that are linked:

```js
// pseudo-code

import { eachDeep } from 'deepdash-es/standalone';

eachDeep(node, (value, key) => {
  if (key.endsWith(`___NODE`)) {
    // can be either single node or array
	if (Array.isArray(value)) {
      value.forEach(linkedNodeId => touchNode({ nodeId: linkedNodeId }))
    } else {
      touchNode({ nodeId: value })
    }
  }
})
```